README.md

```diff
@@ -4,35 +4,97 @@ The web app lets you:
 - create a jazz lyric
 - see the jazz lyric & reports the number of vowels in the lyric
 - see a jazz song comprised of all the previously-created jazz lyrics
-
-View a deployed version at: `http://kimlew.pythonanywhere.com/`
-
-This app:
 - has & uses Flask function decorators
 - uses the Jinja2 text template engine
 
-This app has been tested with:
+See a deployed version at:
+
+`http://kimlew.pythonanywhere.com/`
+
+See the Pipfile for the requirements and dependencies. See the Pipfile.lock for the specific versions. This app has been tested with:
 - Python 3.10
 - Flask 2.2.3
 - Jinja2 3.1.2
 - python-dotenv 1.0.0
 - mysql-connector-python 8.0.32
-- mysql 8.0  
-
-See the Pipfile for the requirements and dependencies. See the Pipfile.lock for the specific versions.
+- mysql 8.0
 
 ### To run the web app locally
 
-1. Clone the repo: `git clone https://github.com/kimlew/pythonapp-createjazzlyric`
+#### A. Make sure you have MySQL set up locally with a new user and password.
+The MySQL installation and setup differs on a Mac, on Linux, etc. Refer to sites like:
+
+`https://database.guide/install-mysql-on-a-mac/`
+
+`https://www.digitalocean.com/community/tutorials/how-to-install-mysql-on-ubuntu-22-04`
+
+
+#### B. Create a Python environment and get the project files
+
+1. Clone the repo:
+  
+  **For PR**:
+
+  `git clone -b python3-10_pipenv https://github.com/kimlew/pythonapp-createjazzlyric`
+
+  **After PR merged**:
 
-2. Change directory: `cd pythonapp-createjazzlyric`
+  `git clone https://github.com/kimlew/pythonapp-createjazzlyric`
+
+1. Change to the project directory with:
+ 
+  `cd pythonapp-createjazzlyric`
+
+3. Create a Python environment with version 3.10:
+ 
+  `python3.10 -m venv .venv`
```

Comment:
I'm not sure I understand why you create both a `venv` environment and also have `pipenv` create another environment below. You never activate the `venv` by running `source ./.venv/bin/activate`, so I'm not sure it's doing anything. Or am I misunderstanding something?

Reply:
It was the only way I could get the environment to recognize Python 3.10 & get the app running vs. using the Mac system Python 3 version of Python 3.9.6, which would NOT run the app.

But, I've edited the steps so it's clear **to run the app locally** and **for deployment to PythonAnywhere**, to use the `venv` module, `pip3` and `requirements.txt`, and to only use *pipenv* during *development*.

Reply:
After doing some experimenting I now understand why you had to do this -- and you were doing the right thing, so sorry for my lack of knowledge.

Apparently, `pipenv` will check to see if there is a `.venv` directory in your project directory. If there is, it will use that `.venv` directory as the virtual environment into which `pipenv install` will install the packages. So creating the `.venv` directory like you did is a way to control where the virtual environment is.

If instead you let `pipenv` create the virtual environment, it will create it in a different directory.

For example, I tried these steps

```
$ python3.10 -m venv .venv
$ pipenv --venv
/Users/kcharter/projects/kim/pythonapp-createjazzlyric/.venv
$ pipenv install
```

you can see with `pipenv --venv` that `pipenv` is going to use the `.venv` I just created (`/Users/kcharter/projects/kim/pythonapp-createjazzlyric` is where I am working).

Then if I do either

```
source .venv/bin/activate
```

or

```
pipenv shell
```

then I can run your app, and `pip3 freeze` shows the right thing.

Then

```
$ pipenv --rm
```

removes the `.venv` directory.

At this point, if I just run `pipenv install --python 3.10` then `pipenv` creates a totally new virtual environment outside the project directory

```
$ pipenv install --python 3.10
...
✔ Successfully created virtual environment!
Virtualenv location: /Users/kcharter/.local/share/virtualenvs/pythonapp-createjazzlyric-gfDUAeGM
Installing dependencies from Pipfile.lock (6e6064)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
```

I can still do `pipenv shell` and then run the application, and `pip3 freeze` produces the right output too, it's just that the virtual env is in `/Users/kcharter/.local/share/virtualenvs/pythonapp-createjazzlyric-gfDUAeGM` instead of in a `.venv` directory in the project.

Reply:
I think I vaguely remember a year or two ago, that pipenv recognizes if you name the environment **.venv** or **venv**, e.g., `python3.10 -m venv .venv`, it creates the environment in a different location than if you use the command virtualenv, as you stated above, but then you can use `pipenv` commands. I couldn't find where I explicitly read that. Maybe I learned it in a LinkedIn tutorial.

Reply:
Yeah, the `pipenv` documentation doesn't mention it, at least not that I can see.

Reply:
@kcharter I've pared down the steps to run the app locally & I only want to show 1 way to do this, since that is simplest. I've stuck to using the way with: `pip3 install -r requirements.txt`

In my Development process, I will make sure the `Pipfile`, `Pipfile.lock` & `requirements.txt` are sync-ed up, so a user can use the `requirements.txt` reliably when running the app locally.

Side Note: When I'm in my environment during Development, this is **cleaner**:
```
(.venv) mac$ pip3 freeze > requirements.txt
(.venv) mac$ cat requirements.txt
click==8.1.3
Flask==2.2.3
itsdangerous==2.1.2
Jinja2==3.1.2
MarkupSafe==2.1.2
mysql-connector-python==8.0.32
protobuf==3.20.3
python-dotenv==1.0.0
Werkzeug==2.2.3
```
**than**:
```
(.venv) mac$ pipenv requirements > requirements.txt
mac$ cat requirements.txt
-i https://pypi.org/simple
click==8.1.3 ; python_version >= '3.7'
flask==2.2.3
itsdangerous==2.1.2 ; python_version >= '3.7'
jinja2==3.1.2
markupsafe==2.1.2 ; python_version >= '3.7'
mysql-connector-python==8.0.32
protobuf==3.20.3 ; python_version >= '3.7'
python-dotenv==1.0.0
werkzeug==2.2.3 ; python_version >= '3.7'
```
For some reason, there is extra text the pipenv way

Reply:
I'm not sure if the extra text is interpreted by `pip3` but perhaps the documentation would say. But it's nice to have the shorter version. Interesting also that the one generated by `pipenv` has alll-lowercase package names.

Reply:
1. So I found out the extra text after some packages in `requirements.txt' created with `pipenv requirements > requirements.txt` are **envionment markers**.
See: `https://peps.python.org/pep-0508/#environment-markers`
```
...
Environment markers allow a dependency specification to provide a rule that describes when the dependency should be used.
...
```
I'm leaving them out since the project ONLY uses Python 3.10.

2. WTR to the packages in uppercase or lowercase, most current examples I see online show uppercase, so I'm sticking with that.